RoaringBitmap/src/main/java/org/roaringbitmap/art/LeafNode.java

```diff
@@ -79,6 +79,11 @@ public int getChildPos(byte k) {
     throw new UnsupportedOperationException();
   }
 
+  @Override
+  public SearchResult getNearestChildPos(byte key) {
```

Comment:
I really like this, I had gone down the path of looping while not "direction greater".

Over all I really like this work, I had some extra tests I had added to  ArtTest.java which all pass with this change, and all my external project tests pass, so thank you so much for that fixes.

my one question is should the getNearestChildPos functions (for each node type) have some tests explicitly written from them. The code looked a little hairy, (aka I've not read each of the node types implementation code). I am happy to write those tests to understand them, and post them on another PR, unless the code is "simple once understood".

Anyways here are the tests I had for testing the seek of the LeafNodeIterator, if you want to include them (or not, or would just like them as another PR).

```

  @Test
  public void testNodeSeek() {
    byte[] key0 = new byte[]{1, 2, 3, 4, 5, 0};
    byte[] key1 = new byte[]{1, 2, 3, 4, 5, 1};
    byte[] key2 = new byte[]{1, 2, 3, 4, 5, 2};
    Art art = new Art();
    insert5PrefixCommonBytesIntoArt(art, 3);
    LeafNodeIterator lnIt = art.leafNodeIterator(false, null);
    boolean hasNext = lnIt.hasNext();

    // brand new iterator is pointing at first key "0"
    Assertions.assertTrue(hasNext);
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the current spot is fine
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the next spot "1" is fine
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    // seeking to the next spot "2" is fine
    lnIt.seek(LongUtils.toLong(key2, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));
    Assertions.assertTrue(BytesUtil.same(lnIt.next().getKeyBytes(), key2));
    Assertions.assertFalse(lnIt.hasNext());

    // seeking to the prior "1" takes you there.. so this needs to be guarded against, in higher levels
    // (as it is)
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));
  }

  @Test
  public void testNodeSeekReverse() {
    byte[] key0 = new byte[]{1, 2, 3, 4, 5, 0};
    byte[] key1 = new byte[]{1, 2, 3, 4, 5, 1};
    byte[] key2 = new byte[]{1, 2, 3, 4, 5, 2};
    Art art = new Art();
    insert5PrefixCommonBytesIntoArt(art, 3);
    LeafNodeIterator lnIt = art.leafNodeIterator(true, null);
    boolean hasNext = lnIt.hasNext();

    // brand new iterator is pointing at first key "2"
    Assertions.assertTrue(hasNext);
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));

    // seeking to the current spot is fine
    lnIt.seek(LongUtils.toLong(key2, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));

    // seeking to the next spot "1" is fine
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    // seeking to the next spot "0" is fine
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
    Assertions.assertTrue(BytesUtil.same(lnIt.next().getKeyBytes(), key0));
    Assertions.assertFalse(lnIt.hasNext());

    // seeking to the prior "1" takes you there.. so this needs to be guarded against, in higher levels
    // (as it is)
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));
  }

  @Test
  public void testNodeSeekOverGaps() {
    byte[] start0 = new byte[]{0, 2, 3, 4, 5, 0};
    byte[] key0 = new byte[]{1, 2, 3, 4, 5, 0};
    byte[] gap1 = new byte[]{1, 2, 3, 4, 5, 1};
    byte[] key1 = new byte[]{1, 2, 3, 4, 5, 2};
    byte[] gap2 = new byte[]{1, 2, 3, 4, 5, 3};
    byte[] key2 = new byte[]{1, 2, 3, 4, 5, 4};
    byte[] end1 = new byte[]{2, 2, 3, 4, 5, 4};

    Art art = new Art();
    insert5PrefixCommonWithGapBytesIntoArt(art, 3);
    LeafNodeIterator lnIt = art.leafNodeIterator(false, null);
    boolean hasNext = lnIt.hasNext();

    // brand new iterator is pointing at first key "0"
    Assertions.assertTrue(hasNext);
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the current spot is fine
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the next spot "1" is fine
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    // seeking to the next spot "2" is fine
    lnIt.seek(LongUtils.toLong(key2, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));
    Assertions.assertTrue(BytesUtil.same(lnIt.next().getKeyBytes(), key2));
    Assertions.assertFalse(lnIt.hasNext());

    // seeking to the prior "1" takes you there.. so this needs to be guarded against, in higher levels
    // (as it is)
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // NOW TO ENTER THE GAP ZONE... we should get "1" as it's after gap "1"
    lnIt.seek(LongUtils.toLong(gap1, (char) 0));

    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    lnIt.seek(LongUtils.toLong(gap2, (char) 0));

    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));

    // going to past the last value should "work", there just should be nothing to get next...
    lnIt.seek(LongUtils.toLong(end1, (char) 0));
    Assertions.assertFalse(lnIt.hasNext());

    // going the before the first should "return" the first
    lnIt.seek(LongUtils.toLong(start0, (char)0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

  }

  @Test
  public void testNodeSeekOverGapsReverse() {
    byte[] start0 = new byte[]{2, 2, 3, 4, 5, 0};
    byte[] key0 = new byte[]{1, 2, 3, 4, 5, 4};
    byte[] gap1 = new byte[]{1, 2, 3, 4, 5, 3};
    byte[] key1 = new byte[]{1, 2, 3, 4, 5, 2};
    byte[] gap2 = new byte[]{1, 2, 3, 4, 5, 1};
    byte[] key2 = new byte[]{1, 2, 3, 4, 5, 0};
    byte[] end1 = new byte[]{0, 2, 3, 4, 5, 0};

    Art art = new Art();
    insert5PrefixCommonWithGapBytesIntoArt(art, 3);
    LeafNodeIterator lnIt = art.leafNodeIterator(true, null);
    boolean hasNext = lnIt.hasNext();

    // brand new iterator is pointing at first key "0"
    Assertions.assertTrue(hasNext);
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the current spot is fine
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // seeking to the next spot "1" is fine
    lnIt.seek(LongUtils.toLong(key1, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    // seeking to the next spot "2" is fine
    lnIt.seek(LongUtils.toLong(key2, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));
    Assertions.assertTrue(BytesUtil.same(lnIt.next().getKeyBytes(), key2));
    Assertions.assertFalse(lnIt.hasNext());

    // seeking to the prior "1" takes you there.. so this needs to be guarded against, in higher levels
    // (as it is)
    lnIt.seek(LongUtils.toLong(key0, (char) 0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));

    // NOW TO ENTER THE GAP ZONE... we should get "1" as it's after gap "1"
    lnIt.seek(LongUtils.toLong(gap1, (char) 0));

    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));

    lnIt.seek(LongUtils.toLong(gap2, (char) 0));

    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
    Assertions.assertFalse(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key1));
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key2));

    // going to past the last value should "work", there just should be nothing to get next...
    lnIt.seek(LongUtils.toLong(end1, (char) 0));
    Assertions.assertFalse(lnIt.hasNext());

    // going the before the first should "return" the first
    lnIt.seek(LongUtils.toLong(start0, (char)0));
    Assertions.assertTrue(lnIt.hasNext());
    Assertions.assertTrue(BytesUtil.same(lnIt.peekNext().getKeyBytes(), key0));
  }
```

Reply:
I think all those extra tests are a great idea. This code is so involved I don't think it can possibly be tested enough.

Would you mind adding anything you can think of (the ones you posted and the per node tests you suggested) in a separate PR? I think it makes more sense for them to come from you instead of me continuing to copy&paste your work into my commits :D Also maybe this PR is monstrous enough as is^^